Update 1.21.11 [DO NOT MERGE - TRACKING]

Dueris opened the pull request:
This is simply for tracking purposes. **DO NOT MERGE**

https://github.com/PaperMC/Folia/tree/ver/1.21.11
https://github.com/CraftCanvasMC/Canvas/tree/ver/1.21.11
## TODO
- [x] release weaver 2.3.11 for AT'd library imports
- [x] cleanup server build file patch (we dont need to override netty now)
- [x] update minecraft base patches
- [x] update paper base patches
- [x] update paper api base/file patches
- [x] update paper file patches
- [x] update minecraft file patches
- [x] moonrise executor rewrite probably useless. needs removal outright or rework -- structure task is parallel safe in canvas, could use this. also need to check parallel capabilities and fix that in the chunk task scheduler
- [x] test world creation/unloading, restore initWorld call in create world
- [x] rework aquifer and beardifier optimizations
- [x] compile fixes
- [x] at separation - this also stops some weird issues from happening when we enforce correct task ordering in weaver
  - as of weaver 2.4.0-SNAPSHOT builds will break on unseparated AT files
- [x] build system updates for new `apiVersion` gradle property
- [x] make RSP command error outputs more specific - https://discord.com/channels/1168986665038127205/1169003058122989620/1451376121990877265
- [x] mobcap issue??? - affects everything except monsters seemingly - should also be fixed on .8
  - https://discord.com/channels/1168986665038127205/1169003058122989620/1451378178877816953 https://discord.com/channels/1168986665038127205/1169003058122989620/1451290560982814731
https://discord.com/channels/1168986665038127205/1169003058122989620/1451260892128874538
  - some additional info:
    - it began sometime recently judging by peoples' reports -> might be related to the scheduler rewrite (could also be related to returnTask optimization that is causing a crash?)

- [x] possible crash caused by build 595 (between 593 and 596) - https://discord.com/channels/1168986665038127205/1169003088389087333/1451295356586430638 -- invalid
- [x] rewrite scheduler queue
- [x] redo repo setup so that base patches can be applied completely solo, and is compilable without file patches applied.
  - [x] should we move our own sources to a different location and then exclude that when doing *just* base patches?
- [x] separate waypoints rewrite and world load/unload API rewrite to base patches

## Future plans
- [x] Support for patch roulette in forks (weaver) - https://github.com/CraftCanvasMC/weaver/commit/ef688aab6cfa66bf92dac61296596387c4cc56d4
- [x] get rid of javadoc mappings - https://github.com/CraftCanvasMC/Canvas/commit/1cbcea2202b958771c3c57474d36851cfc3578b6

Dueris committed:
Promote CI channel to 1.21.11-experimental

Dueris committed:
Update build system

Dueris committed:
Fix apiVersion not set

Dueris committed:
Move all to TODO, begin update process

Dueris committed:
Update API Base Patches

Dueris committed:
Update API File Patches

Dueris committed:
Update Folia profiler removal patch

Dueris committed:
Update Vanilla profiler removal patch

Dueris committed:
Specify in MD file that it is the Folia profiler

Dueris committed:
Update Purpur Ender Chest 6 Rows Config patch

Dueris committed:
Update Disable Criterion Trigger Config patch

Dueris committed:
Update Remove Dead Old Watchdog Code patch

Dueris committed:
Update Rewrite Folia Scheduler patch

Toffikk committed:
chore: remove useless netty override

Toffikk committed:
fix: add back io_uring dep

Dueris committed:
Prepare rework of Moonrise executor replacement

Dueris committed:
Annotate new thread pool replacement

Dueris committed:
Remove old executor thread, replacement

Toffikk committed:
chore: at-ify concurrent util

Toffikk committed:
chore: remove unapplicable ats

Toffikk committed:
chore: dev-import BalancedPrioritisedThreadPool

Dueris committed:
Update Moonrise Executor Rewrite

Toffikk committed:
chore: weaver v2.3.11

Dueris committed:
Update all paper server patches

Dueris committed:
Merge remote-tracking branch 'origin/ver/1.21.11' into ver/1.21.11

Dueris committed:
Move Moonrise patches to work

Dueris committed:
Begin final phase of patch updates

Dueris committed:
Move Aquifer to work

Toffikk committed:
chore: update gradle wrapper file

Toffikk committed:
chore: AT seperation

this speeds up builds by a bit and also future proofs us

Toffikk committed:
chore: test plugin build system updates

Dueris committed:
Apply Minecraft File Patches

Dueris committed:
Slightly adjust placement of 2 patches

Dueris committed:
Merge remote-tracking branch 'origin/ver/1.21.11' into ver/1.21.11

Dueris committed:
All patches applied \o/

Dueris committed:
Fix compile errors

Dueris committed:
Remove log4j changes, fixup buildscript with test plugins

Dueris committed:
Fix startup crash with sword item properties passing

Dueris committed:
Fixup intentionally passing null tick time during runTasks due to recent SpottedLeaf changes causing crashes

Toffikk commented on build.gradle.kts:
```diff
@@ -101,6 +101,7 @@ subprojects {
         apply(plugin = "xyz.jpenilla.resource-factory-paper-convention")
         dependencies {
             compileOnly(rootProject.projects.canvasServer)
+            compileOnly(rootProject.projects.canvasApi)
```
this shouldn’t be needed as the server project bundles the api project

Dueris replied:
From when I was doing compile fixes, my local test plugins failed because it couldn't see API files for some reason. That fixed it

Toffikk committed:
chore: move to AT

Toffikk committed:
chore: move this to ATs too

Toffikk committed:
chore: switch to snapshot weaver for now

Toffikk committed:
fix: don't report build time as EPOCH

Dueris committed:
Replace Moonrise Executor

This generally is more powerful than Moonrises current executor in Paper, and also is a bit faster than 1.21.8 Canvas. Chunk *loading* should also be faster aswell, since with the recent Moonrise changes it allows us to safely replace the IO pool with an optimized varient that won't completely destroy the server.

I generally saw a double in CPS with these changes as the improvement, this is up for more testing though.

Dueris committed:
Pull https://github.com/PaperMC/Folia/pull/412 early

When using the `/ride` command in 1.21.11 on an entity in the same region, it works fine. However, when attempting to ride an entity out of region, it causes a crash due to how the `Entity#startRiding` command is written, making it not region threading safe when attempting to ride an entity out of the current region.

This was tested and is replicatable in the latest Folia 1.21.11 commit by forceloading a chunk(or chunk*s*) far outside the current region, and then running the command:
```
/ride @s mount @e[limit=1,type=!player,sort=furthest]
```

This effectively tries to teleport the mounter -> mount target, where the mount target and mounter are not in the same region. The fix in this PR provides a fix for the entirety of the `Entity#startRiding` method, and also a thread check at the HEAD of the method to ensure that `Entity#startRiding` is called on the region of the mounter entity.

TODO - when/if this is merged, this commit needs to be reverted

Tested-by: Dueris <duerismc@gmail.com>

Toffikk committed:
Merge remote-tracking branch 'origin/ver/1.21.8' into ver/1.21.11

Dueris committed:
Update Upstream (Folia)

This upstream conforms to the scheduler refactoring done by SpottedLeaf, adding a new CRS configuration option for Canvas, which is based off the EDF scheduler but allows processing intermediate tasks.

All changes mimicking Folia's recent upstream(which there were many) have been removed in preference for retaining the upstream version. The default scheduler option will remain EDF, as it has proven reliable. CRS is required for region profiling though, and the CRS scheduler will continue to be improved upon in the future.

Dueris committed:
Remove optimizedAutoSave option

This option has proven to be useful for servers with slow disks, but very bad for performance in other cases. It has been decided to remove this option, as the amount of people with performance decreases far outweighs the people with slow disks having performance increases.

Dueris committed:
Update Upstream (Folia)

Dueris committed:
Remove sending respawn packet on async teleport

Generally, this feature was useful and a good QOL feature for 1.21.8 servers at the time. Now, this packet does something different, and no longer displays the screen that it was intended to display. As such, it has been decided to remove this feature to match with Folia, as I see no point in continuing to contain and maintain this in Canvas.

gbpii committed:
Fix GLOBAL_BROADCAST to only send messages to operators (#151)

- Added operator check in GLOBAL_BROADCAST consumer
- Messages now only appear to ops instead of all players
- Console logging remains unchanged

(cherry picked from commit dd6794d5c60cdd52e6814d3e5abdf120226dde25)

Co-authored-by: gbpii <gbpii@users.noreply.github.com>

Dueris committed:
Reimplement and update Beardifier optimizations

Dueris committed:
Reimplement and update Aquifer optimizations

Dueris committed:
Promote channel to stable

Dueris committed:
Update Upstream (Folia)

Dueris committed:
Literally how did this change not make it in the last commit...

Toffikk committed:
[ci-skip] feat: update for weaver auto AT file discovery for non-minecraft sources (#152)

* feat: update for weaver auto at discovery

* chore: switch to stable weaver

Dueris committed:
Update Upstream (Folia)

Dueris committed:
Adjust how save-all patch works

Dueris committed:
[ci-skip] Add pre-jenkins task to optimize workflow efficiency

Dueris committed:
Improve CRS Scheduler

Dueris committed:
Fix Folia Visibility API

Generally, this API is slow af and just not thread-safe.
With Folia, this is an extreme issue. Due to this API being
not thread-safe, and the way this is called and designed
is completely unsafe. {@link CraftPlayer#onEntityRemove(Entity)}
is called in tracking end of non-player-entities, on *all* players...
This is being called on a non thread-safe field, which thus
causes threading issues from even the visibility API being
used by a plugin, even though it's not the plugins fault.

This is genuinely something we need to call, and we shouldn't
schedule this(because with enough regions and players this
will spam the entity scheduler), and we can't really do much
beyond make this API faster and thread-safe. As such, this
field has been changed to a {@link java.util.concurrent.ConcurrentHashMap}
which has O(1), or worst-case, O(log(n)) time complexity for removals
and lookups, which is sufficient for the purposes we need this for.

While probably not the greatest way to go about this, there
aren't many practical ways to go about resolving this issue,
which is desperately in need of fixing. Large production
servers that I(Dueris) have been in contact with crash every
few hours due to this issue, which isn't even caused by them,
but the fact they use this API in general. There were other ideas
proposed to fix this, yet none of them were entirely practical
or efficient. This is generally the best way to go about this
without completely disabling or rewriting this API from the ground
up with a different architecture and design to work with region threading

This could probably be reworked in the future to be better, but this does
fix the issues caused by the visibility API.

We can guarentee this is safe though, since we are trying to fix entity
*removal*, which normally we can only add/remove visibility invertedness when
on the owning player entities context. The method call we are trying to fix,
when called on the same context, won't cause issues, and the only cross-thread
interaction that can occur is the removal of the entity’s visibility entry
itself on tracking removal, which is now handled through a thread-safe map.
This ensures that entity teardown cannot corrupt internal state, while all visibility
inversion mutations remain confined to the owning player’s context and therefore
preserve the original single-threaded semantics of the API. If a situation where
say the removal and adding of an inverted visibility was to occur(like onEntityRemove
and just adding visibility inverstion), this wouldn't do anything, given the
entity is removed anyway if it *does* get added. If it's a removal and onEntityRemove,
this doesn't matter since they do the same thing ultimately

Dueris committed:
Fix Spark Thread Pattern

During Folia's refactoring of the scheduler system, SpottedLeaf changed the name of the Thread pattern for the scheduler threads. This was unnoticed by myself until a few reports of spark data not showing up in profilers, of which then I found the issue. This commit fixes up the pattern in the PinningThreadDumper, making spark profiler reports accurate to reflect the 1.21.11 scheduler change

Dueris committed:
Pull Current Folia PR Party

Dueris committed:
Restore max concurrent login patch

Dueris committed:
Luminol fix incorrect ticket locking area while calling setSource on propagator in addTicketAtLevel

This fix was discussed by M2 and myself in DMs, seems to fix a critical issue in the propagator in Folia 1.21.11

Dueris committed:
Fix https://github.com/PaperMC/Folia/issues/431

Dueris committed:
Update Upstream (Folia)

Dueris commented:
Finished 1.21.11 update, no need to keep tracking open.
